src/data/error.rs

```diff
@@ -79,3 +80,136 @@ impl<T, E> Recoverable for Result<T, (E, T)> {
         }
     }
 }
+
+pub enum CompileErrorLevel {
+    Warning,
+    RecoverableError,
+    FatalError,
+}
+
+pub trait NewCompileError {
+    // Idealy this should be an associated constant, but those are not allowed
+    // in trait objects
+    fn level(&self) -> CompileErrorLevel;
+
+    // Could also be an associated const
+    fn name(&self) -> &'static str;
+
+    fn message(&self) -> String;
+    fn location(&self) -> Location;
+}
+
+impl fmt::Debug for (dyn NewCompileError + 'static) {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}: {}", self.name(), self.message())
+    }
+}
+
+impl PartialEq for (dyn NewCompileError + 'static) {
+    fn eq(&self, other: &Self) -> bool {
+        self.name() == other.name()
+    }
+}
+
+impl fmt::Display for (dyn NewCompileError + 'static) {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{} ({})", self.message(), self.name())
+    }
+}
+
+macro_rules! define_error {
+    ($error_short_name:expr, $error_level:expr, $error_name:ident($($arg_name:ident: $arg_type:ty ),*), $message:expr) => {
```

Comment:
The usage here is:
```rust
define_error! {
    "short-name", error_level,
    ErrorName(arg_name: ArgType),
    "formatting string"
}
```

There's a little bit of magic in the type name, arguments, and formatting string.